google-cloud-bigtable/lib/google-cloud-bigtable.rb

```diff
@@ -0,0 +1,222 @@
+    #   Provides the means for authenticating requests made by the client. This parameter can
+    #   be many types.
+    #   A `Google::Auth::Credentials` uses a the properties of its represented keyfile for
+    #   authenticating requests made by this client.
+    #   A `String` will be treated as the path to the keyfile to be used for the construction of
+    #   credentials for this client.
+    #   A `Hash` will be treated as the contents of a keyfile to be used for the construction of
+    #   credentials for this client.
+    #   A `GRPC::Core::Channel` will be used to make calls through.
+    #   A `GRPC::Core::ChannelCredentials` for the setting up the RPC client. The channel credentials
+    #   should already be composed with a `GRPC::Core::CallCredentials` object.
+    #   A `Proc` will be used as an updater_proc for the Grpc channel. The proc transforms the
+    #   metadata for requests, generally, to give OAuth credentials.
+    # @param scopes [Array<String>]
+    #   The OAuth scopes for this service. This parameter is ignored if an
+    #   updater_proc is supplied.
+    # @param client_config [Hash]
+    #   A Hash for call options for each method.
+    #   See Google::Gax#construct_settings for the structure of
+    #   this data. Falls back to the default config if not specified
+    #   or the specified config is missing data points.
+    # @param timeout [Integer]
+    #   The default timeout, in seconds, for calls made through this client.
+    # @return [InstanceAdminClient | TableAdminClient | DataClient]
+    # @example Create instance admin client
+    #   require "google/cloud/bigtable"
+    #
+    #   client = Google::Cloud.bigtable(
+    #     project_id: "project-id",
+    #     client_type: :instance
+    #   )
+    #
+    # @example Create table admin client
+    #   require "google/cloud/bigtable"
+    #
+    #   client = Google::Cloud.bigtable(
+    #     project_id: "project-id",
+    #     instance_id: "instance_id"
+    #     client_type: :table
+    #   )
+    #
+    # @example Create table data operations client
+    #   require "google/cloud/bigtable"
+    #
+    #   client = Google::Cloud.bigtable(
+    #     project_id: "project-id",
+    #     instance_id: "instance_id",
+    #     client_type: :data
+    #   )
+
+    def self.bigtable \
+        project_id: nil,
+        instance_id: nil,
+        client_type: :data,
+        credentials: nil,
+        scopes: nil,
+        client_config: nil,
+        timeout: nil
+      gem_spec = Gem.loaded_specs["google-cloud-bigtable"]
+      options = {
+        credentials: credentials,
+        scopes: scopes,
+        client_config: client_config,
+        timeout: timeout,
+        lib_name: gem_spec.name,
+        lib_version: gem_spec.version.to_s
+      }
+
+      raise InvalidArgumentError, "project_id is required" unless project_id
+
+      if client_type == :instance
```

Comment:
This logic should happen in `Google::Cloud::Bigtable.new`.

Are you planning on looking at GCE metadata, environment variables, or adding configuration support for the construction arguments?

Reply:
Got it. Moving code to `Google::Cloud::Bigtable.new`